### PR TITLE
MRG: sklearn api compat

### PR DESCRIPTION
**Pull request by mathurinm**
So that `celer_path` can be used in `CVestimator` of sklearn

### CODE REVIEW

File: celer/tests/test_homotopy.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    X, y, _, _ = build_dataset(n_samples=30, n_features=50, n_targets=1)
+
+    params = dict(eps=1e-2, n_alphas=10, tol=1e-8)
+    alphas1, coefs1, gaps1 = celer_path(X, y,
+                                        return_thetas=False, verbose=False,
+                                        **params)
+
+    alphas2, coefs2, gaps2 = lasso_path(X, y, verbose=False, **params)
+
+    np.testing.assert_allclose(alphas1 / len(X), alphas2)
```
Comment (agramfort) [inline celer/tests/test_homotopy.py]: @mathurinm you already have a mistmath between sklearn and celer.

sklearn solves the pb with 1/n_samples * the loss

now this does not example the random segfault :(

File: celer/tests/test_homotopy.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    tol = 1e-6
+    alphas, coefs, gaps, thetas = celer_path(X, y, alphas=alphas, tol=tol,
+                                             return_thetas=True, verbose=False,
+                                             verbose_inner=False)
+    np.testing.assert_array_less(gaps, tol)
+
+
+def test_celer_path_vs_lasso_path():
+    """Test that our estimator is pluggable into sklearn's LassoCV."""
+    X, y, _, _ = build_dataset(n_samples=30, n_features=50, n_targets=1)
```
Comment (agramfort) [inline celer/tests/test_homotopy.py]: hint : I don't get any segfault if I set n_samples a bit bigger.

**Comment by agramfort** [comment]
@mathurinm I also tried to remplace the np.empty by np.zeros to get a more deterministic result but it still seems random outputs. I don't get it...

**Comment by agramfort** [comment]
@mathurinm we should have proper Lasso and LassoCV classes in celer. That would act as dropin replacement of sklearn

from celer import Lasso, LassoCV

ok?

**Comment by mathurinm** [comment]
sure

File: celer/dropin_sklearn.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        super(Lasso, self).__init__(**kwargs)
+        self.alpha = alpha
+        self.max_iter = max_iter
+        self.gap_freq = gap_freq
+        self.max_epochs = max_epochs
+        self.p0 = p0
+        self.verbose = verbose
+        self.tol = tol
+        self.prune = prune
+        self.fit_intercept = False  # fit intercept by adding a dummy column
```
Comment (agramfort) [inline celer/dropin_sklearn.py]: not true. no dummy column is added in sklearn.

File: celer/dropin_sklearn.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+from .wrapper import celer
+from .homotopy import celer_path
+
+
+class Lasso(Lasso_sklearn):
+    """Drop-in replacement for sklearn.linear_model.Lasso."""
+
+    def __init__(self, alpha=1., max_iter=100, gap_freq=10,
+                 max_epochs=50000, p0=10, verbose=0, tol=1e-6,
+                 prune=0, **kwargs):
```
Comment (agramfort) [inline celer/dropin_sklearn.py]: avoid kwargs and pass parameters explicitly

File: celer/homotopy.py [hunk 5]
```diff
@@ -1,11 +1,18 @@
...
 from .wrapper import celer
 
 
-def celer_path(X, y, alphas=None, max_iter=20, gap_freq=10,
-               max_epochs_inner=50000, p0=10, verbose=1, verbose_inner=1,
-               tol=1e-6, prune=0):
+def celer_path(X, y, eps=1e-3, n_alphas=100, alphas=None, max_iter=20,
+               gap_freq=10, max_epochs=50000, p0=10, verbose=0,
+               verbose_inner=0, tol=1e-6, prune=0, return_thetas=False,
+               **kwargs):
```
[the remark this answers is not in the document]
Comment (mathurinm) [reply celer/homotopy.py]: https://github.com/scikit-learn/scikit-learn/blob/a24c8b46/sklearn/linear_model/coordinate_descent.py#L742

in this call there are parameters like X_offset or precompute that I don't handle so if I don't put them the code breaks

**Comment by agramfort** [comment]
don't cripple the public API. use private functions

File: celer/dropin_sklearn.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class LassoCV(LassoCV_sklearn):
+    """Drop-in replacement for sklearn.linear_model.LassoCV."""
+
+    def __init__(self, eps=1e-3, n_alphas=100, alphas=None,
+                 fit_intercept=False, max_iter=100,
+                 tol=1e-6, cv=None, verbose=0, gap_freq=10,
+                 max_epochs=50000, p0=10, prune=0,
+                 normalize=False, precompute='auto'):
```
Comment (mathurinm) [inline celer/dropin_sklearn.py]: @agramfort if I don't put normalize and precompute, then later on clf.get_params() does not contain 'normalize' nor 'precopute' and clf.fit() fails...
But putting them here is misleading as they are not used

Comment (agramfort) [reply celer/dropin_sklearn.py]: don't put them here. Introduce a private path function with **kwargs but don't cripple the public API

File: examples/plot_lasso_cv.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+The example runs the LassoCV scikit-learn like estimator
+using the Celer algorithm.
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+from sklearn.datasets import fetch_mldata
+
+from celer import LassoCV
```
Comment (agramfort) [inline examples/plot_lasso_cv.py]: I don't get the same figure if I replace this line with:

from sklearn.linear_model import LassoCV

Comment (mathurinm) [reply examples/plot_lasso_cv.py]: Probably fit intercept defaulting to True. I'll fix it and the non ascii characters when I land

**Comment by mathurinm** [comment]
because my way of dealing with intercept was to add a constant column with huge values. Right now there is no way to update an intercept inside the Cython solver.

**Comment by mathurinm** [comment]
I'll have a look

**Comment by mathurinm** [comment]
@agramfort OK to merge on your side?

**Comment by agramfort** [comment]
thx @mathurinm 

to put in a Tweet can you compare the time of LassoCV on Finance between sklearn and celer?

speed up should make a good tag line :)
